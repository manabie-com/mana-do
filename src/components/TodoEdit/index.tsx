import React, { useRef, useEffect } from "react";
import TodoInput from "components/TodoInput";

export interface TodoEditProps {
  value?: string;
  onSubmit: (value: string) => void;
  onCloseEdit: () => void;
}

const TodoEdit: React.FC<TodoEditProps> = ({
  value = "",
  onSubmit,
  onCloseEdit,
}) => {
  const editRef = useRef<HTMLDivElement>(null);
  useEffect(() => {
    document.addEventListener("click", handleClickOutside);
    return () => {
      document.removeEventListener("click", handleClickOutside);
    };
  }, []); // eslint-disable-line react-hooks/exhaustive-deps

  function handleClickOutside(event: MouseEvent) {
    if (editRef.current && !editRef.current.contains(event.target as Node)) {
      onCloseEdit();
    }
  }

  return (
    <div ref={editRef}>
      <TodoInput
        data-testid="todo-input"
        autoFocus={true}
        defaultValue={value}
        onEnter={onSubmit}
      />
    </div>
  );
};

export default TodoEdit;
