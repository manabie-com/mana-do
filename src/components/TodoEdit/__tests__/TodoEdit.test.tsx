import React from "react";
import { expect, test, describe } from "@jest/globals";
import { fireEvent, render, screen } from "@testing-library/react";
import TodoEdit, { TodoEditProps } from "../index";
import { isPropertySignature } from "typescript";

const mockedOnSubmit = jest.fn();
const mockedOnCloseEdit = jest.fn();

const defaultProps: TodoEditProps = {
  value: "",
  onSubmit: mockedOnSubmit,
  onCloseEdit: mockedOnCloseEdit,
};

describe("TodoEdit Snapshots", () => {
  test("should render snapshot with no text", () => {
    const { container } = render(<TodoEdit {...defaultProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test("should render snapshot with some text", () => {
    const props: TodoEditProps = { ...defaultProps, value: "shopping" };
    const { container } = render(<TodoEdit {...props} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("TodoEdit functions", () => {
  test("should render corrent text of value", () => {
    const props: TodoEditProps = { ...defaultProps, value: "shopping" };
    render(<TodoEdit {...props} />);
    const inputElement = screen.getByRole("textbox") as HTMLInputElement;
    expect(inputElement.value).toBe("shopping");
  });

  test("should auto focus the input", () => {
    const props: TodoEditProps = { ...defaultProps, value: "shopping" };
    render(<TodoEdit {...props} />);
    const inputElement = screen.getByRole("textbox") as HTMLInputElement;
    expect(document.activeElement).toEqual(inputElement);
  });

  test("should fire onSubmit event when press Enter", () => {
    const props: TodoEditProps = { ...defaultProps, value: "shopping" };
    render(<TodoEdit {...props} />);
    const inputElement = screen.getByRole("textbox") as HTMLInputElement;
    fireEvent.keyDown(inputElement, { key: "Enter" });
    expect(mockedOnSubmit).toBeCalled();
  });
});
