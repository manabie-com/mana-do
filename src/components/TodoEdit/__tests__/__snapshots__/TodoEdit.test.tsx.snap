// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoEdit Snapshots should render snapshot with no text 1`] = `
<div>
  <TodoInput
    autoFocus={true}
    data-testid="todo-input"
    defaultValue=""
    onEnter={[MockFunction]}
  />
</div>
`;

exports[`TodoEdit Snapshots should render snapshot with some text 1`] = `
<div>
  <input
    class="Todo__input"
    placeholder="What need to be done?"
    value="shopping"
  />
</div>
`;
